packages/dds/tree/src/feature-libraries/modular-schema/schemaChangeFamily.ts

```diff
@@ -0,0 +1,278 @@
+        return this.encodeForJsonI(formatVersion, this.rootSchema, change);
+    }
+
+    private encodeForJsonI(
+        formatVersion: number,
+        schemaId: TreeSchemaIdentifier | undefined,
+        change: NodeChangeset,
+    ): JsonCompatibleReadOnly {
+        const schemaIdFinal = change.schema ?? schemaId;
+        assert(schemaIdFinal !== undefined, "Expected schema ID");
+        const schema = this.getTreeSchema(schemaIdFinal);
+
+        const fields: FieldChangeMap = {};
+        for (const field of Object.keys(change.fields)) {
+            const childSchema = this.getChildSchema(schema, brand(field));
+            fields[field] = this.getChangeHandler(schema, brand(field)).encoder.encodeForJson(
+                formatVersion,
+                change.fields[field],
+                (childChange) => this.encodeForJsonI(formatVersion, childSchema, childChange),
+            );
+        }
+
+        return {
+            ...change,
+            fields,
+        };
+    }
+
+    public encodeBinary(formatVersion: number, change: NodeChangeset): Buffer {
+        throw new Error("Method not implemented.");
+    }
+
+    public decodeJson(formatVersion: number, change: JsonCompatibleReadOnly): NodeChangeset {
+        return this.decodeJsonI(formatVersion, this.rootSchema, change);
+    }
+
+    private decodeJsonI(
+        formatVersion: number,
+        schemaId: TreeSchemaIdentifier | undefined,
+        encoded: JsonCompatibleReadOnly,
+    ): NodeChangeset {
+        const change = encoded as unknown as NodeChangeset;
+        const schemaIdFinal = change.schema ?? schemaId;
+        assert(schemaIdFinal !== undefined, "Expected schema ID");
+        const schema = this.getTreeSchema(schemaIdFinal);
+
+        const fields: FieldChangeMap = {};
+        for (const field of Object.keys(change.fields)) {
+            const childSchema = this.getChildSchema(schema, brand(field));
+            fields[field] = this.getChangeHandler(schema, brand(field)).encoder.decodeJson(
+                formatVersion,
+                change.fields[field],
+                (encodedChild) => this.decodeJsonI(formatVersion, childSchema, encodedChild),
+            );
+        }
+
+        return {
+            ...change,
+            fields,
+        };
+    }
+
+    public decodeBinary(formatVersion: number, change: Buffer): NodeChangeset {
+        throw new Error("Method not implemented.");
+    }
+
+    private getChangeHandler(schema: TreeSchema, field: FieldKey): FieldChangeHandler<any> {
+        // TODO: Check extra fields
+        const fieldSchema = schema.localFields.get(field as LocalFieldKey)
+            ?? this.fieldSchemaForGlobalField(schema, field as GlobalFieldKey);
+
+        assert(fieldSchema !== undefined, "Unknown field key");
+        const fieldKind = this.schemaPolicy.fieldKinds.get(fieldSchema.kind);
+        assert(fieldKind !== undefined, "Unknown field kind");
+        return fieldKind.changeHandler;
+    }
```

Comment:
It's likely that the work of getting the FieldKind for a given FieldKey in a TreeSchema should just be some static helper function in the schema code.

Reply:
Agreed